linkerd/distribute/src/keys.rs

```diff
@@ -0,0 +1,152 @@
+use rand::{
+    distributions::{WeightedError, WeightedIndex},
+    prelude::Distribution as _,
+    rngs::SmallRng,
+};
+use std::{collections::HashMap, hash::Hash};
+
+// Uniquely identifies a key/backend pair for a distribution. This allows
+// backends to have the same key and still participate in request distribution.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct KeyId {
+    idx: usize,
+}
+
+impl KeyId {
+    pub(crate) fn new(idx: usize) -> Self {
+        Self { idx }
+    }
+}
+
+#[derive(Debug)]
+pub struct UnweightedKeys<K> {
+    ids: Vec<KeyId>,
+    keys: HashMap<KeyId, K>,
+}
+
+pub type WeightedKeys<K> = UnweightedKeys<WeightedKey<K>>;
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct WeightedKey<K> {
+    pub key: K,
+    pub weight: u32,
+}
+
+pub(crate) struct WeightedKeySelector<'a, K> {
+    keys: &'a WeightedKeys<K>,
+    index: WeightedIndex<u32>,
+}
+
+// === impl UnweightedKeys ===
+
+// PartialEq, Eq, and Hash are all valid to implement for UnweightedKeys since
+// there is a defined iteration order for the keys, but it cannot be automatically
+// derived for HashMap fields.
+impl<K: PartialEq> PartialEq for UnweightedKeys<K> {
+    fn eq(&self, other: &Self) -> bool {
+        if self.ids != other.ids {
+            return false;
+        }
+
+        for id in &self.ids {
+            if self.keys.get(id) != other.keys.get(id) {
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+impl<K: Eq> Eq for UnweightedKeys<K> {}
+
+impl<K: Hash> Hash for UnweightedKeys<K> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.ids.hash(state);
+        // Normally we would also hash the length, but self.ids and
+        // self.keys have the same length
+        for id in &self.ids {
+            self.keys.get(id).hash(state);
+        }
+    }
+}
+
+impl<K> UnweightedKeys<K> {
+    pub(crate) fn new(iter: impl Iterator<Item = K>) -> Self {
+        let mut ids = Vec::new();
```

Comment:
This is how the invariant of the IDs and keys being a 1:1 relation is guaranteed by only allowing construction via enumeration of the iterator over keys.